scripts/abi_artifact_generator/templates/functions.inc.jinja_template

```diff
@@ -62,3 +62,28 @@ NOINLINE /*( return_value_type )*/ ABIDEF test_/*( name )*/(void) {
 }
 
 /*{ endfor }*/
+
+// To not let the linker get rid of the functions we want to see, we need a way
+// for them to be traceable from main. This generates a function that calls
+// every test.
+#ifdef LINKER_HACK
+void prevent_linker_from_removing_tests(void) {
+/*{ for name, arguments in argument_functions.items() }*/
+  {
+    /*{ for type in arguments -}*/
+    /*( type )*/ argument_/*( loop.index - 1 )*/ = *(/*( type )*/ *) expected_state;
+    /*{ endfor -}*/
+    test_/*( name )*/(
+      /*{- for type in arguments -}*/
+      argument_/*( loop.index - 1 )*/
+      /*{- if not loop.last -}*/ , /*{ endif -}*/
+      /*{- endfor -}*/
+    );
+  }
+/*{- endfor }*/
+
+/*{ for name in return_value_functions -}*/
+  test_/*( name )*/();
+/*{ endfor }*/
+}
+#endif
```

Comment:
1. Still hesitating on it. Once I wrapped up the bugfixes I have open now (since I'm almost done with that PR), I was planning to look deeper into it.

That should help with the tail-call workaround. Not so sure about this one though. Since, the problem here is the msvc linker. And we kind of need `/opt:ref` otherwise the binary is really big (filled with a whole lot of stuff we don't care about) which makes all the stuff we do with it really slow. And `/opt:ref` also happens to remove all these functions unless they are reachable from the main.

Reply:
2. > which makes all the stuff we do with it really slow

   How can the linker make us slower? I'm a bit suspicious. I would understand if, by default, there's a lot of code and `/opt:ref` DCE's it away, but otherwise I don't see how a "bigger binary" can make us slower. Is there more code? Just padding?

Reply:
Don't forget that the binary is linked statically. MSVC toolchain pulls a whole lot of stuff in. `/opt:ref` eliminates practically all of it.

Reply:
Why link statically?

Reply:
Because we call functions from said binary from gcc. It's definitely a lot safer (for a certain definition of safer) for the binary to be self-contained and to not try to dynamically link to anything.

Reply:
I suspect this is FUD: we already rely on the fact that the functions we call do not call anyone else.

Let's avoid complications, and use regular dynamic binaries unless you have proof that we need statically linked binaries.

Reply:
Ping?

Reply:
So what's the final decision on -O0 vs -O2?

I went with `pragma optimize off` around the offender.

Reply:
OK, please add a comment about why we need this.

Are there any other places where we need to do tricks to fight the compiler trying to optimize the code away?
IIRC the argument in favor of `-O2` over `-O0` is readability of the code, but let's make sure we don't need 10 hacks around the codebase to keep `-O2`.

Reply:
It's specifically the tail call optimization. And even then - it's only because of the way we inject break-points. We could, _theoretically_ rework that part instead - with support for tail calls, but, let's be real, it's a lot easier to just force compiler not to optimize those.

Also, it's only `-O0` on MSVC, on GCC it's `#pragma GCC optimize ("no-optimize-sibling-calls")`.

Reply:
IMO, I'd still rather go with the "fake-call" (I also tried various other ideas you had, like writing the address of the function to a global variable - they all still cause trouble to varying degrees) to prevent optimization, but this works.

Reply:
> Also, it's only -O0 on MSVC, on GCC it's #pragma GCC optimize ("no-optimize-sibling-calls").

I'm not concerned about the fact that we disable optimizations. I'm concerned about the fact that we *enable* them.

So my question is: are there other places that would become significantly simpler if we were to globally switch from `-O2` to `-O0`?

Reply:
> are there other places that would become significantly simpler if we were to globally switch from `-O2` to `-O0`?

Not that I know of. Aside from pretty obvious tail-call related trap (we never hit the breakpoint we set at the instruction right after the `call` one), I haven't see anything bad happen because of optimizations.

> I'm concerned about the fact that we _enable_ them.

I feel the opposite way. The fact that even something intrusive as compiler optimizing stuff does not break any of our assumptions makes me feel like we're doing something write. Bonus point: the code we're testing is a lot close to the code we're going to see in the real world too.